File: ts/src/resources/sessions.ts
import { HttpTransport } from "../http.js"

export class SessionsResource {
  constructor(private readonly http: HttpTransport, private readonly base: string) {}

  async list(params?: { directory?: string; workspace?: string }) {
    const url = new URL(`${this.base}/session`)
    if (params?.directory) url.searchParams.set("directory", encodeURIComponent(params.directory))
    if (params?.workspace) url.searchParams.set("workspace", params.workspace)
    return this.http.getFetch()(url, { method: "GET" }).then(res => res.json())
  }

  async get(sessionID: string, params?: { directory?: string; workspace?: string }) {
    const url = new URL(`${this.base}/session/${sessionID}`)
    if (params?.directory) url.searchParams.set("directory", encodeURIComponent(params.directory))
    if (params?.workspace) url.searchParams.set("workspace", params.workspace)
    return this.http.getFetch()(url, { method: "GET" }).then(res => res.json())
  }

  async create(body: { directory?: string; workspace?: string }) {
    return this.http.getFetch()(`${this.base}/session`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(res => res.json())
  }

  async delete(sessionID: string) {
    return this.http.getFetch()(`${this.base}/session/${sessionID}`, { method: "DELETE" })
      .then(res => res.json())
  }
}

File: py/src/imbrace/resources/sessions.py
from typing import Optional, Any
from ..http import HttpTransport, AsyncHttpTransport


class SessionsResource:
    """Sessions domain — Sync. 1 file = 1 domain."""
    def __init__(self, http: HttpTransport, base: str):
        self._http = http
        self._base = base

    def list(self, directory: Optional[str] = None, workspace: Optional[str] = None) -> Any:
        params = {}
        if directory:
            params["directory"] = directory
        if workspace:
            params["workspace"] = workspace
        return self._http.request("GET", f"{self._base}/session", params=params).json()

    def get(self, session_id: str, directory: Optional[str] = None) -> Any:
        params = {"directory": directory} if directory else {}
        return self._http.request("GET", f"{self._base}/session/{session_id}", params=params).json()

    def create(self, directory: Optional[str] = None, workspace: Optional[str] = None) -> Any:
        body: dict = {}
        if directory:
            body["directory"] = directory
        if workspace:
            body["workspace"] = workspace
        return self._http.request("POST", f"{self._base}/session", json=body).json()

    def delete(self, session_id: str) -> Any:
        return self._http.request("DELETE", f"{self._base}/session/{session_id}").json()


class AsyncSessionsResource:
    """Sessions domain — Async. 1 file = 1 domain."""
    def __init__(self, http: AsyncHttpTransport, base: str):
        self._http = http
        self._base = base

    async def list(self, directory: Optional[str] = None, workspace: Optional[str] = None) -> Any:
        params = {}
        if directory:
            params["directory"] = directory
        if workspace:
            params["workspace"] = workspace
        res = await self._http.request("GET", f"{self._base}/session", params=params)
        return res.json()

    async def get(self, session_id: str, directory: Optional[str] = None) -> Any:
        params = {"directory": directory} if directory else {}
        res = await self._http.request("GET", f"{self._base}/session/{session_id}", params=params)
        return res.json()

    async def create(self, directory: Optional[str] = None, workspace: Optional[str] = None) -> Any:
        body: dict = {}
        if directory:
            body["directory"] = directory
        if workspace:
            body["workspace"] = workspace
        res = await self._http.request("POST", f"{self._base}/session", json=body)
        return res.json()

    async def delete(self, session_id: str) -> Any:
        res = await self._http.request("DELETE", f"{self._base}/session/{session_id}")
        return res.json()
